### DISCUSSION TITLE
Create generic action providers?

### DISCUSSION THREAD

**Discussion by mattbrailsford** [nathanwoulfe/Plumber-2#120] [Ideas]
We've had a number of requests for our Vendr product where users would love to be able to create custom workflows that are triggered by orders changing status etc and whilst we could develop such a feature, I do wonder if it wouldn't just be a massive duplication of effort given Plumber exists.

From what I can tell with Plumber though, it does look like it's pretty tightly tied to Umbraco content / document types as being the source of actions for a workflow. I was wondering then, how much effort would it be for triggers / action providers to be totally pluggable such that different sources could be added and totally custom workflows for things other than content could be created?

The same could go for actions that could be performed too as for example, in Vendr we have custom email templates so it would be great if as part of a workflow you could trigger a Vendr Email Template Send action.

Would this be at all possible with Plumber? and would it be of interest to achieve? I'd really rather push folks to buy a Plumber license to do these advanced things than for us to have to replicate and maintain a similar package for our needs.

What do you think?

**Comment by nathanwoulfe** [comment]
Hey Matt, just a really quick reply to acknowledge I'm thinking about how this could work. 

You're correct in that Plumber is (super) tightly coupled to content, so that would be a big hurdle to overcome. Plumber is more focussed on approval chains rather than performing actions (or, the action is at the end of the chain, and is a publish or unpublish event). 

Will ponder this some more. If you have a more detailed example of the sort of actions you're thinking, that might be useful too...

**Reply by mattbrailsford** [reply]
Yea, it's a huge change but could potentially open a lot of possibilities.

I guess for Vendr it could be a lot of things, but a few examples could be

1, When an order status changes to "Disptached" send a dispatch email to the customer
2. If an order value exceeds $1,000 tag the order with "vip"
3. If a cart is abandoned for more than 24 hours, email a discount code to the customer
4. If a product becomes back in stock, send an email to customers who signed up to be notified
5. If a customizable order comes in, get approval of the customization before proceeding. There could be a whole management flow here to get updates, and then if successful, the order status moves to "approved".

There are a whole heap of possibilities. A good example of a workflow system from Shopify could be Shopify Flow https://apps.shopify.com/flow which I think could be done generically enough to be useful for a lot of scenarios.

I do think that a flexible workflow package could be useful for a lot of other packages too so could be a nice way to cross pollinate. 

Happy to brainstorm some time and even potentially help in it's creation if it's something you'd be interested in?

**Comment by nathanwoulfe** [comment]
More thinking... It's almost similar to how plugins work in Preflight - the plugin must implement an interface, Preflight then discovers these on startup and queues the plugins to run as part of the content checks.

For Plumber, a similar pattern would be to define a plugin (aka a workflow activity) which has a trigger and an action. The trigger could be an Umbraco notification, a Vendr notification, etc, then the action would do something and return a result (to enable chaining). Vendr (for example would need to provide triggers and actions, then Plumber would provide the interface for linking triggers to actions, and a mechanism for storing and executing. Something like that...

Another example, Plumber could provide a trigger based on Umbraco's content publishing handler, and an action for pinging a Slack channel, so the action would need to accept parameters for interacting with Slack.

I'd see this all initially sitting alongside the existing content-based workflows, which could in time be updated to use the plugin-driven workflow.
